doc/vlmc.1
.\" Copyright (C) 2008-10 VideoLAN
.\" Author: Rohit Yadav <rohityadav89@gmail.com>
.\"
.\" This is free software; you may redistribute it and/or modify
.\" it under the terms of the GNU General Public License as
.\" published by the Free Software Foundation; either version 2,
.\" or (at your option) any later version.
.\"
.\" This is distributed in the hope that it will be useful, but
.\" WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public License 
.\" along with this; if not write to the Free Software Foundation, Inc.
.\" 59 Temple Place, Suite 330, Boston, MA 02111-1307  USA
.TH VLMC 1 "Oct 4, 2010"
.SH NAME
vlmc \- VideoLAN Movie Creator
.SH SYNOPSIS
.B vlmc
.RI [ OPTIONS ]...
.RI [ filename|URI ]...
.SH DESCRIPTION
VideoLAN Movie Creator (VLMC) is a cross-platform, non-linear video editing
software based on libVLC and running on Windows, Linux and Mac OS X.
It is a free software distributed under the terms of the GPL v2.
This manual page documents briefly the
.B VLMC
video editor.
.SH OPTIONS
.TP
.B \-\-help|-?
Outputs help.
.TP
.B \-\-project|\-p [filename|URI]
Specifies the file to open when VLMC starts.
If this is not specified, VLMC will load a blank file with an "Untitled Project" label.
.TP
.B \-\-version
Outputs version information and exits.
.SH FILES
.I ~/.config/vlmc/vlmc.conf
.RS
The configuration file of the VLMC.
.SH BUGS
If you find a bug, please report it at http://trac.videolan.org/vlmc/newticket?type=defect
.PP
.SH AUTHOR
This manual page was written by Rohit Yadav <rohityadav89@gmail.com>, for the Debian
GNU/Linux system (but may be used by others).


NEWS
Changes between 0.1.1 and 0.2.0:

Interface:
* Do not display video informations when an audio file is displayed
* Adding a filter by name/tags feature
* Adding a toolbar
* Import media by dragging them to the media library
* Adding a volume slider

General:
* Adding video filters
* Adding Youtube export
* Projects are now stored in a workspace
* Fixed crash (deadlock) with very short videos.
* Fixed many other crashes.

-------------------------------------------------------------------------------
Changes between 0.1.0 and 0.1.1:

Timeline:
* Fixed clip deletion problems
* The confirmation for clip deletion can be deactivated.
* Fixed trim/resize bug at maximum zoom

Interface:
* Corrected next/previous frame icons
* Corrected many typo/untranslated strings
* Fix a marker refresh bug
* Add status tips on some widgets
* New menu accelerators
* Ask for confirmation when deleting a clip used in the timeline

General:
* Do not consider VLMC crashed last time when launched the first time
* Auto-detect the user's locale at startup
* Save link between clips when saving a project
* Fixed crash when importing a hierarchy of 2+ directories
* Adding missing extensions

Installer:
* Associates *.vlmc files on Windows
* Add desktop shortcut on Windows


src/Main/vlmc.cpp
/*****************************************************************************
 * vlmc.cpp: VLMC launcher for unix
 *****************************************************************************
 * Copyright (C) 2008-2010 VideoLAN
 *
 * Authors: Hugo Beauzee-Luyssen <hugo@vlmc.org>
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston MA 02110-1301, USA.
 *****************************************************************************/

#include "config.h"

#include "Types.h"

#include <QtDebug>
#include <QMetaType>
#include <QTextStream>
#include <QVariant>

#include <sys/wait.h>
#include <unistd.h>


int VLMCmain( int , char** );

#if defined(WITH_CRASHHANDLER) && defined(Q_OS_UNIX)

#ifdef WITH_GUI
    #include "project/GuiProjectManager.h"
    #ifdef WITH_CRASHHANDLER_GUI
        #include "CrashHandler.h"
    #endif
#else
    #include "ProjectManager.h"
#endif

void
signalHandler( int sig )
{
    signal( sig, SIG_DFL );

    #ifdef WITH_GUI
        GUIProjectManager::getInstance()->emergencyBackup();
    #else
        ProjectManager::getInstance()->emergencyBackup();
    #endif

    #ifdef WITH_CRASHHANDLER_GUI
        CrashHandler* ch = new CrashHandler( sig );
        ::exit( ch->exec() );
    #else
        ::exit( 1 );
    #endif
}
#endif

/**
 * Print version text
 */
void
version( void )
{
    QTextStream out( stdout );
    out << "VLMC-" << PROJECT_VERSION << " '" << CODENAME << "'\n"
        << "VideoLAN Movie Creator (VLMC) is a cross-platform, non-linear\n"
        << "video editing software.\n"
        << "Copyright (C) 2008-10 VideoLAN\n"
        << "This is free software; see the source for copying conditions. There is NO\n"
        << "warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.\n";
}

/**
 * Print usage text
 */
void
usage( QString const& appName )
{
    QTextStream out( stdout );
    out << "Usage: " << appName << " [options] [filename|URI]...\n"
        << "Options:\n"
        << "\t[--project|-p projectfile]\tload the given VLMC project\n"
        << "\t[--version]\tversion information\n"
        << "\t[--help|-?]\tthis text\n\n"
        << "\tFILES:\n"
        << "\t\tFiles specified on the command line should include \n"
        << "\t\tVLMC project files (.vlmc)\n";
}

int
main( int argc, char **argv )
{
    /* Check for command arguments */
    for( int i = 1; i < argc; i++ )
    {
        QString arg = argv[i];

        if( arg == "--help" || arg == "-?" )
        {
            ::usage( QString(argv[0]) );
            return 2;
        }
        else if( arg == "--version" )
        {
            ::version();
            return 2;
        }
    }

    #ifdef WITH_CRASHHANDLER
        while( true )
        {
            pid_t pid = fork();
            if( pid < 0 )
                qFatal("Can't fork to launch VLMC. Exiting.");
            if( pid == 0 )
            {
                signal( SIGSEGV, signalHandler );
                signal( SIGFPE, signalHandler );
                signal( SIGABRT, signalHandler );
                signal( SIGILL, signalHandler );
                return VLMCmain( argc, argv );
            }
            else
            {
                int status;

                wait( &status );
                if( WIFEXITED(status) )
                {
                    int ret = WEXITSTATUS( status );
                    if ( ret == 2 )
                        continue ;
                    else
                        break ;
                }
                else
                {
                    qCritical() << "Unhandled crash.";
                    break ;
                }
            }
        }
    #else
        return VLMCmain( argc, argv );
    #endif
}
